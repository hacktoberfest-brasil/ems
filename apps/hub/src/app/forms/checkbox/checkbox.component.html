<!-- <div
  class="mat-slide-toggle mat-slide-toggle-label-before"
  *ngIf="vm$ | async as vm"
  [class.mat-checked]="vm.checked"
>
  <label class="mat-slide-toggle-label" #label>
    <div
      #toggleBar
      class="mat-slide-toggle-bar"
      [class.mat-slide-toggle-bar-no-side-margin]="
        !labelContent.textContent || !labelContent.textContent.trim()
      "
    >
      <input
        #input
        class="mat-slide-toggle-input cdk-visually-hidden"
        type="checkbox"
        role="switch"
        [checked]="vm.checked"
        [attr.aria-checked]="vm.checked.toString()"
        (change)="onChangeEvent($event)"
      />

      <div class="mat-slide-toggle-thumb-container" #thumbContainer>
        <div class="mat-slide-toggle-thumb"></div>
        <div
          class="mat-slide-toggle-ripple mat-focus-indicator"
          mat-ripple
          [matRippleTrigger]="label"
          [matRippleCentered]="true"
          [matRippleRadius]="20"
          [matRippleAnimation]="{ enterDuration: 150 }"
        >
          <div
            class="mat-ripple-element mat-slide-toggle-persistent-ripple"
          ></div>
        </div>
      </div>
    </div>

    <span class="mat-slide-toggle-content" #labelContent>
      <span style="display: none;">&nbsp;</span>
      <ng-content></ng-content>
    </span>
  </label>
</div> -->
  <div *ngIf="vm$ | async as vm" class="checkbox">
    <input
      id="b1"
      class="cdk-visually-hidden"
      name="checks"
      type="checkbox"
      #input
      [checked]="vm.checked"
      [attr.aria-checked]="vm.checked.valueOf()"
      (change)="onChangeEvent($event)"
    />
    <label for="b1">
      <ng-content></ng-content>

      <pre>{{ vm$ | async | json }}</pre>

      <svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg">
        <path
          d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
        />
      </svg>
    </label>
  </div>
