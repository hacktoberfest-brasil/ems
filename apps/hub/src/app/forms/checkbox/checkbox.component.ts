import {
  Component,
  Input,
  ChangeDetectionStrategy,
  ElementRef,
  Output,
  ViewChild,
  ViewEncapsulation,
} from '@angular/core';
import { ComponentStore } from '@ngrx/component-store';
import { tap } from 'rxjs/operators';

export interface CheckboxToggleState {
  checked: boolean;
}

/** Change event object emitted by a CheckboxToggleComponent. */
export interface CheckboxToggleChange {
  /** The source CheckboxToggle of the event. */
  readonly source: CheckboxComponent;
  /** The new `checked` value of the CheckboxToggle. */
  readonly checked: boolean;
}

let seekId = 0;

@Component({
  selector: 'emx-checkbox',
  templateUrl: './checkbox.component.html',
  styleUrls: ['./checkbox.component.scss'],
  // encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [ComponentStore],
})
export class CheckboxComponent {
  private _id = `emx-checkbox-${seekId++}`;
  public get id() {
    return this._id;
  }
  @Input()
  public set id(value) {
    this._id = value;
  }
  @Input() set checked(value: boolean) {
    this.setChecked(value);
  }
  // Observable<MatSlideToggleChange> used instead of EventEmitter
  @ViewChild('input') inputElement: ElementRef<HTMLInputElement>;

  readonly setChecked = this.componentStore.updater(
    (state, value: boolean) => ({ ...state, checked: value })
  );

  // ViewModel for the component
  readonly vm$ = this.componentStore.select((state) => ({
    checked: state.checked,
  }));

  constructor(
    private readonly componentStore: ComponentStore<CheckboxToggleState>
  ) {
    // set defaults
    this.componentStore.setState({
      checked: false,
    });
  }

  onChangeEvent = this.componentStore.effect<Event>((event$) => {
    return event$.pipe(
      tap<Event>((event) => {
        event.stopPropagation();
        this.setChecked(this.inputElement.nativeElement.checked);
      })
    );
  });
}
