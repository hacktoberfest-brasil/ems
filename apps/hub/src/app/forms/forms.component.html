<h1>forms</h1>

<!-- <form [formGroup]="form">

  <emx-checkbox>
    Another checkbox
  </emx-checkbox>

</form> -->
