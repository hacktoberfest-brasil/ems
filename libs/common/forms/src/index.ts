export * from './lib/checkbox';
export * from './lib/radio';
export * from './lib/submit';
