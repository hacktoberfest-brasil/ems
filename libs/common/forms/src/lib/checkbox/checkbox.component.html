<div class="emx-checkbox">
  <input
    [id]="id"
    class="cdk-visually-hidden"
    name="checks"
    type="checkbox"
    #input
    [value]="value ? value : false"
    [checked]="vm?.checked"
    [attr.aria-checked]="vm?.checked"
    [formControl]="ngControl?.control"
    (change)="onChangeEvent($event)"
  />
  <label [attr.for]="id">
    <ng-content></ng-content>

    <svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg">
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
