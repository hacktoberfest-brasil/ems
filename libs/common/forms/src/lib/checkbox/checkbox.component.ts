import {
  Self,
  Input,
  Output,
  Optional,
  ViewChild,
  Component,
  Injectable,
  ElementRef,
  forwardRef,
  OnDestroy,
  EventEmitter,
  AfterViewInit,
  AfterContentInit,
  ChangeDetectionStrategy,
} from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { EventInputTarget } from './../types';
import {
  AbstractControl,
  FormControl,
  NgControl,
  NG_VALUE_ACCESSOR,
} from '@angular/forms';
import { EmxControlAccessor } from '../abstractions';

let nextId = 0;

@Injectable()
export class EmxCheckbox extends EmxControlAccessor {
  @Input()
  public set disabled(value: boolean) {
    this._disabled = value;
  }
  @Input()
  public set value(value: any) {
    this._value = value;
  }
}

const EmxCheckboxProvider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => EmxCheckbox),
  multi: true,
};

@Component({
  selector: 'emx-checkbox',
  templateUrl: './checkbox.component.html',
  styleUrls: ['./checkbox.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [EmxCheckbox, EmxCheckboxProvider],
})
export class EmxCheckboxComponent extends EmxCheckbox
  implements AfterContentInit, AfterViewInit, OnDestroy {
  destroy$ = new Subject<void>();

  @ViewChild('input', { static: true }) _el: ElementRef<HTMLInputElement>;
  get el() {
    return this._el.nativeElement;
  }

  @Input()
  public set id(value: string) {
    this._id = value;
  }
  public get id(): string {
    return this._id;
  }
  private _id = `emx-checkbox-${nextId++}`;

  @Output()
  valueChange: EventEmitter<any> = new EventEmitter<any>();

  @Output()
  checkedChange = new EventEmitter<boolean>(false);

  control: AbstractControl;

  constructor(@Optional() @Self() public ngControl: NgControl) {
    super();
  }

  ngAfterContentInit() {
    this.control = this.ngControl?.control
      ? this.ngControl.control
      : new FormControl();
  }
  ngAfterViewInit() {
    this.ngControl?.control?.valueChanges
      .pipe(takeUntil(this.destroy$))
      .subscribe((value) => this.valueChange.emit(value));
  }

  onChangeEvent(evt: EventInputTarget) {
    this.onChange(evt.target.value);
    this.checkedChange.emit(evt.target.checked);
    this.valueChange.emit(this.control.value);
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }
}
