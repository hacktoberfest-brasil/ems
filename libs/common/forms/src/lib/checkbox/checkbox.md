## Import

```ts
import { EmxCheckboxModule } from '@emx/components';
```


