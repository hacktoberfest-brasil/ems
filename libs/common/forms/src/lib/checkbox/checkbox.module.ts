import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { EmxCheckboxComponent } from './checkbox.component';

export * from './checkbox.component';

@NgModule({
  imports: [
    FormsModule,
    CommonModule,
    ReactiveFormsModule
  ],
  declarations: [EmxCheckboxComponent],
  exports: [EmxCheckboxComponent]
})
export class EmxCheckboxModule { }
