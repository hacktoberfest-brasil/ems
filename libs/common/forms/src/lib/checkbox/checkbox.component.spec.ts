import { Component } from '@angular/core';
import { FormsModule, ReactiveFormsModule, FormControl } from '@angular/forms';
import {
  Spectator,
  SpectatorHost,
  createHostFactory,
  createComponentFactory,
} from '@ngneat/spectator';

import { EmxCheckboxComponent } from './checkbox.component';

const VIEWS = {
  onlyCheck: `<emx-checkbox [formControl]="form"></emx-checkbox>`,
  withLabel: `<emx-checkbox [formControl]="form">Label</emx-checkbox>`,
  withCustomId: `<emx-checkbox id="x" [formControl]="form"></emx-checkbox>`,
};

@Component({ selector: 'emx-custom-form', template: '' })
class CustomFormComponent {
  form = new FormControl();
}

describe('EmxCheckboxComponent', () => {
  describe('Host', () => {
    let spectator: SpectatorHost<EmxCheckboxComponent, CustomFormComponent>;
    const createHost = createHostFactory({
      component: EmxCheckboxComponent,
      host: CustomFormComponent,
      imports: [ReactiveFormsModule],
    });

    describe('Views', () => {
      it('should display only input', () => {
        spectator = createHost(VIEWS.onlyCheck);
        expect(spectator.query('.emx-checkbox')).toMatchSnapshot();
      });

      it('should display checkbox with label', () => {
        spectator = createHost(VIEWS.withLabel);
        expect(spectator.query('.emx-checkbox')).toMatchSnapshot();
      });

      it('should display checkbox with custom id', () => {
        spectator = createHost(VIEWS.withCustomId);
        expect(spectator.query('.emx-checkbox')).toMatchSnapshot();
      });
    });

    describe('Props', () => {
      it('should create an instance', () => {
        spectator = createHost(VIEWS.withCustomId);
        const input = spectator.query('.emx-checkbox > input');
        expect(input.getAttribute('id')).toEqual('x');
      });
    });
  });

  describe('Forms', () => {
    let spectator: Spectator<EmxCheckboxComponent>;
    const createComponent = createComponentFactory({
      component: EmxCheckboxComponent,
      imports: [FormsModule, ReactiveFormsModule],
    });

    it('should create', () => {
      spectator = createComponent();
      expect(spectator.component).toBeTruthy();
    });

    it('should value changed if set value', () => {
      spectator.component.control.setValue('abc');
      expect(spectator.component.control.value).toBe('abc');
    });

    it('should value changed if set value', async () => {
      spyOn(spectator.component.valueChange, 'emit');
      spectator.component.control.valueChanges.subscribe(() => {
        expect(spectator.component.onChangeEvent).toBeCalledTimes(1);
        expect(spectator.component.valueChange.emit).toBeCalledTimes(1);
        expect(spectator.component.checkedChange.emit).toBeCalledTimes(1);
      });
      spectator.component.control.setValue('qwe');
    });
  });
});

