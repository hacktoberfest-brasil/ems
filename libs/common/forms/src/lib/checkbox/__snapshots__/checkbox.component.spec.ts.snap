// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmxCheckboxComponent Host Views should display checkbox with custom id 1`] = `
<div
  class="emx-checkbox"
>
  <input
    aria-checked="false"
    class="cdk-visually-hidden ng-untouched ng-pristine ng-valid"
    id="x"
    name="checks"
    type="checkbox"
    value="false"
  />
  <label
    for="x"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;

exports[`EmxCheckboxComponent Host Views should display checkbox with label 1`] = `
<div
  class="emx-checkbox"
>
  <input
    aria-checked="false"
    class="cdk-visually-hidden ng-untouched ng-pristine ng-valid"
    id="emx-checkbox-1"
    name="checks"
    type="checkbox"
    value="false"
  />
  <label
    for="emx-checkbox-1"
  >
    Label
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;

exports[`EmxCheckboxComponent Host Views should display only input 1`] = `
<div
  class="emx-checkbox"
>
  <input
    aria-checked="false"
    class="cdk-visually-hidden ng-untouched ng-pristine ng-valid"
    id="emx-checkbox-0"
    name="checks"
    type="checkbox"
    value="false"
  />
  <label
    for="emx-checkbox-0"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;
